Crashes when pipelines use list syntax for environment variables

Pipelines can list env vars as lists, similar to docker-compose.

```yaml
steps:
  - command: true
    env:
      - MY_VAR=value
```

This causes the pipeline signing to fail because it expects a map:

https://github.com/seek-oss/buildkite-signed-pipeline/blob/b598adbed132529869ed0f1167a0c524f697c8bd/cmd/buildkite-signed-pipeline/signer.go#L134-L140

```

panic: reflect: call of reflect.Value.MapKeys on slice Value
--
  |  
  | goroutine 1 [running]:
  | reflect.flag.mustBe(0x97, 0x15)
  | /usr/local/Cellar/go/1.11.1/libexec/src/reflect/value.go:207 +0xb4
  | reflect.Value.MapKeys(0x850340, 0xc00049c220, 0x97, 0x3, 0xc0002fa528, 0x1)
  | /usr/local/Cellar/go/1.11.1/libexec/src/reflect/value.go:1133 +0x49
  | main.SharedSecretSigner.signStep(0xc00046c080, 0x20, 0x0, 0x0, 0x876e60, 0xc00049c250, 0x194, 0x1, 0xc0001ae600, 0x0, ...)
  | /Users/zsims/projects/buildkite-signed-pipeline/signer.go:137 +0x50e
  | main.SharedSecretSigner.Sign(0xc00046c080, 0x20, 0x0, 0x0, 0x878960, 0xc0001ae0f0, 0x7f742a184000, 0x0, 0xc000066ba0, 0x0)
  | /Users/zsims/projects/buildkite-signed-pipeline/signer.go:57 +0x668
  | main.(*uploadCommand).run(0xc00000ae40, 0xc00016bc20, 0x859000, 0xc000035450)
  | /Users/zsims/projects/buildkite-signed-pipeline/main.go:100 +0xfa
  | main.(*uploadCommand).run-fm(0xc00016bc20, 0x409af3, 0x88d720)
  | /Users/zsims/projects/buildkite-signed-pipeline/main.go:38 +0x34
  | gopkg.in/alecthomas/kingpin%2ev2.(*actionMixin).applyActions(0xc0000c20d8, 0xc00016bc20, 0x0, 0x0)
  | /Users/zsims/go/pkg/mod/gopkg.in/alecthomas/kingpin.v2@v2.2.6/actions.go:28 +0x6d
  | gopkg.in/alecthomas/kingpin%2ev2.(*Application).applyActions(0xc0000c4690, 0xc00016bc20, 0x0, 0x0)
  | /Users/zsims/go/pkg/mod/gopkg.in/alecthomas/kingpin.v2@v2.2.6/app.go:557 +0xdf
  | gopkg.in/alecthomas/kingpin%2ev2.(*Application).execute(0xc0000c4690, 0xc00016bc20, 0xc000035440, 0x1, 0x1, 0x0, 0x0, 0xc00016e960, 0x0)
  | /Users/zsims/go/pkg/mod/gopkg.in/alecthomas/kingpin.v2@v2.2.6/app.go:390 +0x8f
  | gopkg.in/alecthomas/kingpin%2ev2.(*Application).Parse(0xc0000c4690, 0xc00001c090, 0x3, 0x3, 0x1, 0xc00000c1e8, 0x0, 0x1)
  | /Users/zsims/go/pkg/mod/gopkg.in/alecthomas/kingpin.v2@v2.2.6/app.go:222 +0x1fa
  | main.main()
  | /Users/zsims/projects/buildkite-signed-pipeline/main.go:80 +0xa6a


```

--- a/Makefile
+++ b/Makefile
@@ -15,6 +15,10 @@ dist/buildkite-signed-pipeline-windows-amd64.exe: $(FILES)
 dist/buildkite-signed-pipeline-darwin-amd64: $(FILES)
 	GOOS=darwin GOARCH=amd64 go build -v -ldflags="$(FLAGS)" -o $@ ./cmd/buildkite-signed-pipeline
 
+.PHONY: test
+test:
+	go test -v $(FILES)
+
 .PHONY: clean
 clean:
 	rm -rf dist/

--- a/cmd/buildkite-signed-pipeline/signer.go
+++ b/cmd/buildkite-signed-pipeline/signer.go
@@ -72,6 +72,32 @@ func (s SharedSecretSigner) Sign(pipeline interface{}) (interface{}, error) {
 	return copy.Interface(), nil
 }
 
+func addSignature(env interface{}, signature Signature) (interface{}, error) {
+	// if there's no env, default to the map format
+	if env == nil {
+		env = make(map[string]interface{})
+	}
+
+	switch i := env.(type) {
+	// key=value environment variables
+	case []interface{}:
+		envCopy := make([]interface{}, len(i))
+		copy(envCopy, i)
+		envCopy = append(envCopy, fmt.Sprintf("%s=%s", stepSignatureEnv, signature))
+		return envCopy, nil
+	// map of environment variables
+	case map[string]interface{}:
+		envCopy := make(map[string]interface{})
+		reflectedEnv := reflect.ValueOf(i)
+		for _, key := range reflectedEnv.MapKeys() {
+			envCopy[key.String()] = reflectedEnv.MapIndex(key).Interface()
+		}
+		envCopy[stepSignatureEnv] = signature
+		return envCopy, nil
+	}
+	return nil, fmt.Errorf("Unknown environment type %T", env)
+}
+
 func (s SharedSecretSigner) signStep(step reflect.Value) (interface{}, error) {
 	original := step.Elem()
 
@@ -130,17 +156,10 @@ func (s SharedSecretSigner) signStep(step reflect.Value) (interface{}, error) {
 		return nil, err
 	}
 
-	env := make(map[string]interface{})
-	existingEnv, hasEnv := copy["env"]
-	if hasEnv {
-		reflectedEnv := reflect.ValueOf(existingEnv)
-		for _, key := range reflectedEnv.MapKeys() {
-			env[key.String()] = reflectedEnv.MapIndex(key).Interface()
-		}
+	existingEnv, _ := copy["env"]
+	if copy["env"], err = addSignature(existingEnv, signature); err != nil {
+		return nil, err
 	}
-
-	env[stepSignatureEnv] = signature
-	copy["env"] = env
 
 	return copy, nil
 }

--- a/cmd/buildkite-signed-pipeline/signer_test.go
+++ b/cmd/buildkite-signed-pipeline/signer_test.go
@@ -106,6 +106,11 @@ func TestPipelines(t *testing.T) {
 			`{"steps":[{"command":"echo Hello \"Fred\"","env":{"EXISTING":"existing-value","STEP_SIGNATURE":"signature(echo Hello \"Fred\",)"}}]}`,
 		},
 		{
+			"Command with existing env list",
+			`{"steps":[{"command":"echo Hello \"Fred\"", "env":["EXISTING=existing-value"]}]}`,
+			`{"steps":[{"command":"echo Hello \"Fred\"","env":["EXISTING=existing-value","STEP_SIGNATURE=signature(echo Hello \"Fred\",)"]}]}`,
+		},
+		{
 			"Pipeline with multiple commands",
 			`{"steps":[{"command":["echo Hello World", "echo Foo Bar"]}]}`,
 			`{"steps":[{"command":["echo Hello World","echo Foo Bar"],"env":{"STEP_SIGNATURE":"signature(echo Hello World\necho Foo Bar,)"}}]}`,

